File: controllers/customer.controller.js
/**
 * BarberEase – Customer Controller
 * 
 * Handles customer profile updates, password changes,
 * and admin listings of customers.
 */

const customerService = require('../services/customer.service');

/** GET /api/customers/profile – Get customer profile */
async function getProfile(req, res) {
  const customer = await customerService.getProfile(req.user.id);
  res.json({
    success: true,
    data: customer
  });
}

/** PUT /api/customers/profile – Update customer profile */
async function updateProfile(req, res) {
  const customer = await customerService.updateProfile(req.user.id, req.body);
  res.json({
    success: true,
    message: 'Profile updated successfully',
    data: customer
  });
}

/** PUT /api/customers/profile/password – Change password */
async function changePassword(req, res) {
  const { currentPassword, newPassword } = req.body;
  const result = await customerService.changePassword(req.user.id, currentPassword, newPassword);
  res.json({
    success: true,
    message: result.message
  });
}

/** GET /api/customers – Get all customers (admin) */
async function getAll(req, res) {
  const customers = await customerService.getAllCustomers();
  res.json({
    success: true,
    data: customers
  });
}

/** PUT /api/customers/:id/toggle – Toggle customer status (admin) */
async function toggleStatus(req, res) {
  const customer = await customerService.toggleCustomerStatus(req.params.id);
  res.json({
    success: true,
    message: `Customer account ${customer.is_active ? 'activated' : 'deactivated'}`,
    data: customer
  });
}

module.exports = {
  getProfile,
  updateProfile,
  changePassword,
  getAll,
  toggleStatus
};

/** PUT /api/customers/profile/wallet – Top up wallet balance */
async function topUp(req, res) {
  const { amount } = req.body;
  const customer = await customerService.topUpWallet(req.user.id, amount);

  // Broadcast wallet changes if needed, or send response
  res.json({
    success: true,
    message: `Successfully added $${parseFloat(amount).toFixed(2)} to your wallet balance.`,
    data: customer
  });
}

module.exports = {
  getProfile,
  updateProfile,
  changePassword,
  getAll,
  toggleStatus,
  topUp
};

File: src/main/java/com/barberease/controllers/CustomerController.java
package com.barberease.controllers;

import com.barberease.models.Customer;
import com.barberease.security.UserPrincipal;
import com.barberease.services.CustomerService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/customers")
public class CustomerController {

    @Autowired
    private CustomerService customerService;

    private Long getAuthenticatedCustomerId() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        if (auth == null || !(auth.getPrincipal() instanceof UserPrincipal)) {
            throw new IllegalArgumentException("Authentication required");
        }
        return ((UserPrincipal) auth.getPrincipal()).getId();
    }

    @GetMapping("/profile")
    public ResponseEntity<Map<String, Object>> getProfile() {
        Long customerId = getAuthenticatedCustomerId();
        Customer customer = customerService.getProfile(customerId);

        Map<String, Object> body = new HashMap<>();
        body.put("success", true);
        body.put("data", customer);
        return ResponseEntity.ok(body);
    }

    @PutMapping("/profile")
    public ResponseEntity<Map<String, Object>> updateProfile(@RequestBody Customer data) {
        Long customerId = getAuthenticatedCustomerId();
        Customer customer = customerService.updateProfile(customerId, data);

        Map<String, Object> body = new HashMap<>();
        body.put("success", true);
        body.put("data", customer);
        return ResponseEntity.ok(body);
    }

    @PutMapping("/profile/password")
    public ResponseEntity<Map<String, Object>> changePassword(@RequestBody Map<String, String> request) {
        Long customerId = getAuthenticatedCustomerId();
        String currentPassword = request.get("currentPassword");
        String newPassword = request.get("newPassword");

        customerService.changePassword(customerId, currentPassword, newPassword);

        Map<String, Object> body = new HashMap<>();
        body.put("success", true);
        body.put("message", "Password changed successfully");
        return ResponseEntity.ok(body);
    }

    @GetMapping
    public ResponseEntity<Map<String, Object>> getAll() {
        List<Customer> customers = customerService.getAllCustomers();

        Map<String, Object> body = new HashMap<>();
        body.put("success", true);
        body.put("data", customers);
        return ResponseEntity.ok(body);
    }

    @PutMapping("/{id}/toggle")
    public ResponseEntity<Map<String, Object>> toggleStatus(@PathVariable Long id) {
        Customer customer = customerService.toggleCustomerStatus(id);

        Map<String, Object> body = new HashMap<>();
        body.put("success", true);
        body.put("message", "Customer status updated");
        body.put("data", customer);
        return ResponseEntity.ok(body);
    }

    @PutMapping("/profile/wallet")
    public ResponseEntity<Map<String, Object>> topUpWallet(@RequestBody Map<String, Object> request) {
        Long customerId = getAuthenticatedCustomerId();
        
        Object amountObj = request.get("amount");
        BigDecimal amount = BigDecimal.ZERO;
        if (amountObj != null) {
            if (amountObj instanceof Number) {
                amount = new BigDecimal(amountObj.toString());
            } else if (amountObj instanceof String) {
                amount = new BigDecimal((String) amountObj);
            }
        }

        Customer customer = customerService.topUpWallet(customerId, amount);

        Map<String, Object> body = new HashMap<>();
        body.put("success", true);
        body.put("data", customer);
        return ResponseEntity.ok(body);
    }
}
